File: java/src/RagEngine.java
/**
 * RAG Engine — Java
 * Ingest documents, embed via Anthropic, store in Qdrant (REST API), retrieve context.
 *
 * pom.xml dependencies:
 *   com.squareup.okhttp3:okhttp:4.12.0
 *   com.fasterxml.jackson.core:jackson-databind:2.17.0
 *
 * export ANTHROPIC_API_KEY=sk-ant-...
 * export QDRANT_URL=http://localhost:6333
 *
 * javac -cp okhttp*.jar:jackson*.jar RagEngine.java
 * java  -cp .:okhttp*.jar:jackson*.jar RagEngine
 */

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.node.*;
import okhttp3.*;

import java.security.MessageDigest;
import java.util.*;
import java.util.concurrent.TimeUnit;
import java.util.stream.Collectors;

public class RagEngine {

    // ── Constants ─────────────────────────────────────────────────────────────
    static final String QDRANT_URL    = System.getenv().getOrDefault("QDRANT_URL", "http://localhost:6333");
    static final String ANTHROPIC_KEY = System.getenv("ANTHROPIC_API_KEY");
    static final String EMBED_URL     = "https://api.anthropic.com/v1/embeddings";
    static final String EMBED_MODEL   = "voyage-3";
    static final int    EMBEDDING_DIM = 1024;
    static final String COLLECTION_PREFIX = "rag_";
    static final int    DEFAULT_TOP_K = 5;
    static final int    CHUNK_SIZE    = 800;
    static final int    CHUNK_OVERLAP = 120;

    static final ObjectMapper JSON = new ObjectMapper();
    static final OkHttpClient HTTP = new OkHttpClient.Builder()
        .connectTimeout(30, TimeUnit.SECONDS)
        .readTimeout(60, TimeUnit.SECONDS)
        .writeTimeout(30, TimeUnit.SECONDS)
        .build();
    static final MediaType JSON_MEDIA = MediaType.get("application/json; charset=utf-8");

    // ── Models ─────────────────────────────────────────────────────────────────
    public static class Document {
        public String content;
        public Map<String, String> metadata;
        public String docId;

        public Document(String content, Map<String, String> metadata) {
            this.content  = content;
            this.metadata = metadata;
            this.docId    = sha256Hex(content);
        }
    }

    public static class Chunk {
        String text;
        String docId;
        int chunkIdx;
        Map<String, String> metadata;
        String chunkId;

        Chunk(String text, String docId, int chunkIdx, Map<String, String> metadata) {
            this.text = text;
            this.docId = docId;
            this.chunkIdx = chunkIdx;
            this.metadata = metadata;
            this.chunkId = sha256Hex(docId + ":" + chunkIdx);
        }
    }

    public static class RetrievedChunk {
        public String text;
        public float score;
        public String docId;
        public int chunkIdx;
        public Map<String, String> metadata;

        RetrievedChunk(String text, float score, String docId, int chunkIdx, Map<String, String> metadata) {
            this.text = text;
            this.score = score;
            this.docId = docId;
            this.chunkIdx = chunkIdx;
            this.metadata = metadata;
        }
    }

    // ── Helpers ───────────────────────────────────────────────────────────────
    static String sha256Hex(String s) {
        try {
            MessageDigest md = MessageDigest.getInstance("SHA-256");
            byte[] digest = md.digest(s.getBytes("UTF-8"));
            StringBuilder sb = new StringBuilder();
            for (byte b : digest) sb.append(String.format("%02x", b));
            return sb.substring(0, 16);
        } catch (Exception e) {
            throw new RuntimeException(e);
        }
    }

    static long sha256ToLong(String s) {
        try {
            MessageDigest md = MessageDigest.getInstance("SHA-256");
            byte[] digest = md.digest(s.getBytes("UTF-8"));
            long val = 0;
            for (int i = 0; i < 8; i++) {
                val = (val << 8) | (digest[i] & 0xFF);
            }
            return val & Long.MAX_VALUE;
        } catch (Exception e) {
            throw new RuntimeException(e);
        }
    }

    static List<String> chunkText(String text) {
        List<String> chunks = new ArrayList<>();
        int start = 0;
        while (start < text.length()) {
            int end = Math.min(start + CHUNK_SIZE, text.length());
            String chunk = text.substring(start, end);
            int nextStart = end;
            if (end < text.length()) {
                for (String sep : new String[]{"\n\n", "\n", ". ", " "}) {
                    int pos = chunk.lastIndexOf(sep);
                    if (pos > CHUNK_SIZE / 2) {
                        chunk = text.substring(start, start + pos + sep.length());
                        nextStart = start + pos + sep.length();
                        break;
                    }
                }
            }
            String trimmed = chunk.trim();
            if (!trimmed.isEmpty()) chunks.add(trimmed);
            if (nextStart <= start) break;
            start = Math.max(nextStart - CHUNK_OVERLAP, nextStart == start ? start + 1 : start);
            start = nextStart - CHUNK_OVERLAP;
            if (start < 0) start = 0;
            if (nextStart >= text.length()) break;
        }
        return chunks;
    }

    /** Deterministic pseudo-embedding for offline testing. */
    static float[] pseudoEmbed(String text) {
        try {
            MessageDigest md = MessageDigest.getInstance("SHA-256");
            byte[] digest = md.digest(text.getBytes("UTF-8"));
            float[] vec = new float[EMBEDDING_DIM];
            double norm = 0;
            for (int i = 0; i < EMBEDDING_DIM; i++) {
                float v = (digest[i % digest.length] & 0xFF) / 255.0f * 2 - 1;
                vec[i] = v;
                norm += v * v;
            }
            norm = Math.sqrt(norm);
            if (norm == 0) norm = 1;
            for (int i = 0; i < EMBEDDING_DIM; i++) vec[i] = (float) (vec[i] / norm);
            return vec;
        } catch (Exception e) {
            throw new RuntimeException(e);
        }
    }

    /** Embed a batch of texts. Falls back to pseudo-embedding on API error. */
    static List<float[]> embedTexts(List<String> texts) throws Exception {
        if (ANTHROPIC_KEY == null || ANTHROPIC_KEY.isEmpty()) {
            return texts.stream().map(RagEngine::pseudoEmbed).collect(Collectors.toList());
        }
        ObjectNode body = JSON.createObjectNode();
        body.put("model", EMBED_MODEL);
        ArrayNode inputArr = JSON.createArrayNode();
        texts.forEach(inputArr::add);
        body.set("input", inputArr);

        Request request = new Request.Builder()
            .url(EMBED_URL)
            .header("x-api-key", ANTHROPIC_KEY)
            .header("anthropic-version", "2023-06-01")
            .header("anthropic-beta", "embeddings-2025-01-01")
            .header("content-type", "application/json")
            .post(RequestBody.create(JSON.writeValueAsBytes(body), JSON_MEDIA))
            .build();

        try (Response response = HTTP.newCall(request).execute()) {
            if (!response.isSuccessful()) {
                return texts.stream().map(RagEngine::pseudoEmbed).collect(Collectors.toList());
            }
            JsonNode root = JSON.readTree(response.body().string());
            List<float[]> result = new ArrayList<>();
            for (JsonNode item : root.get("data")) {
                JsonNode embArr = item.get("embedding");
                float[] vec = new float[embArr.size()];
                for (int i = 0; i < embArr.size(); i++) vec[i] = (float) embArr.get(i).asDouble();
                result.add(vec);
            }
            return result;
        } catch (Exception e) {
            return texts.stream().map(RagEngine::pseudoEmbed).collect(Collectors.toList());
        }
    }

    // ── RAG Engine ────────────────────────────────────────────────────────────
    final String collectionName;

    public RagEngine(String name) throws Exception {
        this.collectionName = COLLECTION_PREFIX + name;
        ensureCollection();
    }

    void ensureCollection() throws Exception {
        // Check if collection exists
        Request getReq = new Request.Builder()
            .url(QDRANT_URL + "/collections/" + collectionName)
            .get().build();
        try (Response resp = HTTP.newCall(getReq).execute()) {
            if (resp.code() == 200) return;
        }
        // Create collection
        ObjectNode body = JSON.createObjectNode();
        ObjectNode vectors = JSON.createObjectNode();
        vectors.put("size", EMBEDDING_DIM);
        vectors.put("distance", "Cosine");
        body.set("vectors", vectors);

        Request putReq = new Request.Builder()
            .url(QDRANT_URL + "/collections/" + collectionName)
            .put(RequestBody.create(JSON.writeValueAsBytes(body), JSON_MEDIA))
            .build();
        try (Response resp = HTTP.newCall(putReq).execute()) {
            if (!resp.isSuccessful()) {
                throw new RuntimeException("Failed to create collection: " + resp.code() + " " + resp.body().string());
            }
            System.err.println("[rag] Collection created: " + collectionName);
        }
    }

    public int ingest(List<Document> documents) throws Exception {
        List<Chunk> allChunks = new ArrayList<>();
        for (Document doc : documents) {
            int idx = 0;
            for (String text : chunkText(doc.content)) {
                Map<String, String> meta = new HashMap<>(doc.metadata);
                meta.put("source_doc_id", doc.docId);
                allChunks.add(new Chunk(text, doc.docId, idx++, meta));
            }
        }

        int batchSize = 32;
        for (int start = 0; start < allChunks.size(); start += batchSize) {
            int end = Math.min(start + batchSize, allChunks.size());
            List<Chunk> batch = allChunks.subList(start, end);
            List<String> texts = batch.stream().map(c -> c.text).collect(Collectors.toList());
            List<float[]> vecs = embedTexts(texts);

            ArrayNode points = JSON.createArrayNode();
            for (int i = 0; i < batch.size(); i++) {
                Chunk c = batch.get(i);
                ObjectNode point = JSON.createObjectNode();
                point.put("id", sha256ToLong(c.chunkId));
                ArrayNode vecArr = JSON.createArrayNode();
                for (float v : vecs.get(i)) vecArr.add(v);
                point.set("vector", vecArr);

                ObjectNode payload = JSON.createObjectNode();
                payload.put("text", c.text);
                payload.put("doc_id", c.docId);
                payload.put("chunk_idx", c.chunkIdx);
                payload.put("metadata", JSON.writeValueAsString(c.metadata));
                point.set("payload", payload);
                points.add(point);
            }

            ObjectNode body = JSON.createObjectNode();
            body.set("points", points);

            Request req = new Request.Builder()
                .url(QDRANT_URL + "/collections/" + collectionName + "/points?wait=true")
                .put(RequestBody.create(JSON.writeValueAsBytes(body), JSON_MEDIA))
                .build();
            try (Response resp = HTTP.newCall(req).execute()) {
                if (!resp.isSuccessful()) {
                    throw new RuntimeException("Upsert failed: " + resp.code() + " " + resp.body().string());
                }
            }
        }
        return allChunks.size();
    }

    public List<RetrievedChunk> retrieve(String query, int topK) throws Exception {
        float[] vec = embedTexts(Collections.singletonList(query)).get(0);

        ObjectNode body = JSON.createObjectNode();
        ArrayNode vecArr = JSON.createArrayNode();
        for (float v : vec) vecArr.add(v);
        body.set("vector", vecArr);
        body.put("limit", topK);
        body.put("with_payload", true);

        Request req = new Request.Builder()
            .url(QDRANT_URL + "/collections/" + collectionName + "/points/search")
            .post(RequestBody.create(JSON.writeValueAsBytes(body), JSON_MEDIA))
            .build();

        List<RetrievedChunk> results = new ArrayList<>();
        try (Response resp = HTTP.newCall(req).execute()) {
            if (!resp.isSuccessful()) {
                throw new RuntimeException("Search failed: " + resp.code() + " " + resp.body().string());
            }
            JsonNode root = JSON.readTree(resp.body().string());
            for (JsonNode r : root.get("result")) {
                JsonNode payload = r.get("payload");
                Map<String, String> meta = new HashMap<>();
                if (payload.has("metadata")) {
                    JsonNode metaNode = JSON.readTree(payload.get("metadata").asText("{}"));
                    metaNode.fields().forEachRemaining(e -> meta.put(e.getKey(), e.getValue().asText()));
                }
                results.add(new RetrievedChunk(
                    payload.get("text").asText(),
                    (float) r.get("score").asDouble(),
                    payload.get("doc_id").asText(),
                    payload.get("chunk_idx").asInt(),
                    meta
                ));
            }
        }
        return results;
    }

    public String retrieveAsContext(String query, int topK) throws Exception {
        List<RetrievedChunk> chunks = retrieve(query, topK);
        if (chunks.isEmpty()) return "No relevant context found.";
        StringBuilder sb = new StringBuilder();
        for (int i = 0; i < chunks.size(); i++) {
            RetrievedChunk c = chunks.get(i);
            sb.append(String.format("[%d] (score=%.3f) %s", i + 1, c.score, c.text));
            if (i < chunks.size() - 1) sb.append("\n\n");
        }
        return sb.toString();
    }

    // ── Tool factory for agent loops ─────────────────────────────────────────
    public interface RagToolFn {
        String search(String query, int topK) throws Exception;
    }

    public RagToolFn asTool() {
        return this::retrieveAsContext;
    }

    // ── Demo main ─────────────────────────────────────────────────────────────
    public static void main(String[] args) throws Exception {
        RagEngine engine = new RagEngine("demo");

        List<Document> docs = Arrays.asList(
            new Document("Java is a class-based, object-oriented language designed for portability via the JVM.",
                Map.of("source", "java_docs")),
            new Document("RAG retrieves relevant chunks from a vector database and prepends them to the LLM prompt to ground answers in real data.",
                Map.of("source", "rag_survey")),
            new Document("Qdrant exposes both a gRPC and a REST API and supports cosine, dot, and Euclidean distance metrics.",
                Map.of("source", "qdrant_docs"))
        );

        int count = engine.ingest(docs);
        System.out.println("Ingested " + docs.size() + " docs → " + count + " chunks");

        String ctx = engine.retrieveAsContext("What is RAG?", DEFAULT_TOP_K);
        System.out.println("\nContext for 'What is RAG?':\n" + ctx);
    }
}

File: rust/src/rag_engine.rs
// RAG Engine — Rust
// Ingest documents, embed, store in Qdrant, retrieve context for agent loops.
//
// Add to Cargo.toml:
//   [dependencies]
//   qdrant-client = "1.13"
//   reqwest = { version = "0.12", features = ["json"] }
//   serde = { version = "1", features = ["derive"] }
//   serde_json = "1"
//   sha2 = "0.10"
//   tokio = { version = "1", features = ["full"] }
//   anyhow = "1"
//   hex = "0.4"
//
// export ANTHROPIC_API_KEY=sk-ant-...
// export QDRANT_URL=http://localhost:6333

use std::collections::HashMap;
use std::env;

use anyhow::{Context, Result};
use qdrant_client::prelude::*;
use qdrant_client::qdrant::{
    CreateCollectionBuilder, Distance, SearchPointsBuilder, UpsertPointsBuilder, VectorsConfigBuilder,
    VectorParamsBuilder, PointStruct, Value, vectors_config, vectors,
};
use reqwest::Client as HttpClient;
use serde::{Deserialize, Serialize};
use sha2::{Digest, Sha256};

// ── Config ─────────────────────────────────────────────────────────────────────
const EMBEDDING_DIM: u64 = 1024;
const COLLECTION_PREFIX: &str = "rag_";
pub const DEFAULT_TOP_K: u64 = 5;
const CHUNK_SIZE: usize = 800;
const CHUNK_OVERLAP: usize = 120;

// ── Models ─────────────────────────────────────────────────────────────────────
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Document {
    pub content:  String,
    pub metadata: HashMap<String, String>,
    pub doc_id:   Option<String>,
}

#[derive(Debug, Clone)]
pub struct Chunk {
    pub text:      String,
    pub doc_id:    String,
    pub chunk_idx: usize,
    pub metadata:  HashMap<String, String>,
    pub chunk_id:  String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct RetrievedChunk {
    pub text:      String,
    pub score:     f32,
    pub doc_id:    String,
    pub chunk_idx: usize,
    pub metadata:  HashMap<String, String>,
}

// ── Helpers ────────────────────────────────────────────────────────────────────
fn sha256_hex(s: &str) -> String {
    let mut hasher = Sha256::new();
    hasher.update(s.as_bytes());
    hex::encode(hasher.finalize())[..16].to_string()
}

pub fn chunk_text(text: &str) -> Vec<String> {
    let mut chunks = Vec::new();
    let mut start = 0usize;
    let bytes = text.as_bytes();
    while start < bytes.len() {
        let end = (start + CHUNK_SIZE).min(bytes.len());
        // Find valid UTF-8 boundary
        let mut boundary = end;
        while boundary > start && !text.is_char_boundary(boundary) {
            boundary -= 1;
        }
        let mut chunk = &text[start..boundary];
        let mut next_start = boundary;
        if boundary < bytes.len() {
            for sep in ["\n\n", "\n", ". ", " "] {
                if let Some(pos) = chunk.rfind(sep) {
                    if pos > CHUNK_SIZE / 2 {
                        chunk = &text[start..start + pos + sep.len()];
                        next_start = start + pos + sep.len();
                        break;
                    }
                }
            }
        }
        let trimmed = chunk.trim().to_string();
        if !trimmed.is_empty() {
            chunks.push(trimmed);
        }
        if next_start <= start { break; }
        start = next_start.saturating_sub(CHUNK_OVERLAP);
    }
    chunks
}

// Deterministic pseudo-embedding for testing.
fn pseudo_embed(texts: &[String]) -> Vec<Vec<f32>> {
    texts.iter().map(|text| {
        let mut hasher = Sha256::new();
        hasher.update(text.as_bytes());
        let digest = hasher.finalize();
        let mut vec: Vec<f32> = (0..EMBEDDING_DIM as usize)
            .map(|i| (digest[i % digest.len()] as f32) / 255.0 * 2.0 - 1.0)
            .collect();
        let norm = vec.iter().map(|v| v * v).sum::<f32>().sqrt().max(1e-10);
        vec.iter_mut().for_each(|v| *v /= norm);
        vec
    }).collect()
}

// ── Anthropic embedding request ───────────────────────────────────────────────
#[derive(Serialize)]
struct EmbedRequest {
    model: String,
    input: Vec<String>,
}

#[derive(Deserialize)]
struct EmbedResponse {
    data: Vec<EmbedItem>,
}

#[derive(Deserialize)]
struct EmbedItem {
    embedding: Vec<f32>,
}

async fn embed_texts_api(http: &HttpClient, texts: Vec<String>) -> Result<Vec<Vec<f32>>> {
    let api_key = env::var("ANTHROPIC_API_KEY").context("ANTHROPIC_API_KEY not set")?;
    let resp = http
        .post("https://api.anthropic.com/v1/embeddings")
        .header("x-api-key", &api_key)
        .header("anthropic-version", "2023-06-01")
        .header("anthropic-beta", "embeddings-2025-01-01")
        .json(&EmbedRequest { model: "voyage-3".to_string(), input: texts.clone() })
        .send()
        .await;

    match resp {
        Ok(r) if r.status().is_success() => {
            let body: EmbedResponse = r.json().await?;
            Ok(body.data.into_iter().map(|i| i.embedding).collect())
        }
        _ => {
            // Fallback to pseudo-embedding
            Ok(pseudo_embed(&texts))
        }
    }
}

// ── RAG Engine ─────────────────────────────────────────────────────────────────
pub struct RAGEngine {
    collection_name: String,
    qdrant:          QdrantClient,
    http:            HttpClient,
}

impl RAGEngine {
    pub async fn new(name: &str) -> Result<Self> {
        let url = env::var("QDRANT_URL").unwrap_or_else(|_| "http://localhost:6333".to_string());
        let config = QdrantClientConfig::from_url(&url);
        let qdrant = QdrantClient::new(Some(config))?;
        let engine = Self {
            collection_name: format!("{}{}", COLLECTION_PREFIX, name),
            qdrant,
            http: HttpClient::new(),
        };
        engine.ensure_collection().await?;
        Ok(engine)
    }

    async fn ensure_collection(&self) -> Result<()> {
        if !self.qdrant.collection_exists(&self.collection_name).await? {
            self.qdrant
                .create_collection(&CreateCollectionBuilder::new(&self.collection_name)
                    .vectors_config(VectorsConfigBuilder::new().params_map(
                        [("".to_string(), VectorParamsBuilder::new(EMBEDDING_DIM, Distance::Cosine).build())]
                    )).build())
                .await
                .context("create_collection")?;
            eprintln!("[rag] Collection created: {}", self.collection_name);
        }
        Ok(())
    }

    pub async fn ingest(&self, documents: Vec<Document>) -> Result<usize> {
        let mut all_chunks: Vec<Chunk> = Vec::new();
        for doc in &documents {
            let id = doc.doc_id.clone().unwrap_or_else(|| sha256_hex(&doc.content));
            for (i, text) in chunk_text(&doc.content).into_iter().enumerate() {
                let mut meta = doc.metadata.clone();
                meta.insert("source_doc_id".to_string(), id.clone());
                let chunk_id = sha256_hex(&format!("{}:{}", id, i));
                all_chunks.push(Chunk { text, doc_id: id.clone(), chunk_idx: i, metadata: meta, chunk_id });
            }
        }

        let batch_size = 32;
        for batch in all_chunks.chunks(batch_size) {
            let texts: Vec<String> = batch.iter().map(|c| c.text.clone()).collect();
            let vecs = embed_texts_api(&self.http, texts).await?;
            let points: Vec<PointStruct> = batch.iter().zip(vecs.iter()).map(|(c, vec)| {
                let id_bytes = c.chunk_id.as_bytes();
                let mut num_id = 0u64;
                for b in &id_bytes[..8.min(id_bytes.len())] {
                    num_id = num_id.wrapping_mul(256).wrapping_add(*b as u64);
                }
                let meta_json = serde_json::to_string(&c.metadata).unwrap_or_default();
                PointStruct::new(
                    num_id,
                    vec.clone(),
                    [
                        ("text".to_string(),      Value::from(c.text.as_str())),
                        ("doc_id".to_string(),     Value::from(c.doc_id.as_str())),
                        ("chunk_idx".to_string(),  Value::from(c.chunk_idx as i64)),
                        ("metadata".to_string(),   Value::from(meta_json.as_str())),
                    ].into_iter().collect::<HashMap<_, _>>(),
                )
            }).collect();

            self.qdrant
                .upsert_points(&UpsertPointsBuilder::new(&self.collection_name, points).build())
                .await
                .context("upsert_points")?;
        }
        Ok(all_chunks.len())
    }

    pub async fn retrieve(&self, query: &str, top_k: u64) -> Result<Vec<RetrievedChunk>> {
        let vecs = embed_texts_api(&self.http, vec![query.to_string()]).await?;
        let results = self.qdrant
            .search_points(&SearchPointsBuilder::new(
                &self.collection_name, vecs[0].clone(), top_k,
            ).with_payload(true).build())
            .await
            .context("search_points")?;

        Ok(results.result.into_iter().map(|r| {
            let text      = r.payload.get("text").and_then(|v| v.as_str()).unwrap_or_default().to_string();
            let doc_id    = r.payload.get("doc_id").and_then(|v| v.as_str()).unwrap_or_default().to_string();
            let chunk_idx = r.payload.get("chunk_idx").and_then(|v| v.as_integer()).unwrap_or(0) as usize;
            let metadata  = r.payload.get("metadata")
                .and_then(|v| serde_json::from_str(v.as_str().unwrap_or("{}")).ok())
                .unwrap_or_default();
            RetrievedChunk { text, score: r.score, doc_id, chunk_idx, metadata }
        }).collect())
    }

    pub async fn retrieve_as_context(&self, query: &str, top_k: u64) -> Result<String> {
        let chunks = self.retrieve(query, top_k).await?;
        if chunks.is_empty() {
            return Ok("No relevant context found.".to_string());
        }
        Ok(chunks.iter().enumerate()
            .map(|(i, c)| format!("[{}] (score={:.3}) {}", i + 1, c.score, c.text))
            .collect::<Vec<_>>()
            .join("\n\n"))
    }
}
